From f3807958ba3c9abeb9165e6b831ef341c6b8532b Mon Sep 17 00:00:00 2001
From: sunwenxiang <sunwenxiang133@gmail.com>
Date: Tue, 15 Sep 2026 17:27:14 +0800
Subject: [PATCH] Add sophgo custom features to bmcweb

- CPLD info (CPUA/CPUB/CPUO/Other/PSU/EFuse power states)
- CPLD firmware version via D-Bus xyz.openbmc_project.State.CPLD
- SOL port switching (port0-3 -> CPU0UART0/CPU0UART1/CPU1UART0/CPU1UART1)
- SOL log export (/sol/logs reads /var/log/obmc-console.log)
- Fan speed control via sophgo.fanControl ControlMode
- 1684x board temperature display (AiCard_Temp)
- CSP unsafe-eval for WebAssembly
- Mask asdbg user from account service
- LED identify via GPIO instead of LED GroupManager

Upstream-Status: Inappropriate [sophgo specific]
---
 include/security_headers.hpp         |   2 +-
 redfish-core/lib/account_service.hpp |   2 +-
 redfish-core/lib/led.hpp             |   6 +-
 redfish-core/lib/systems.hpp         | 223 +++++++++++++++++++++++++++
 4 files changed, 228 insertions(+), 5 deletions(-)

diff --git a/include/security_headers.hpp b/include/security_headers.hpp
index 2a2eb40d..1228d32f 100644
--- a/include/security_headers.hpp
+++ b/include/security_headers.hpp
@@ -65,7 +65,7 @@ inline void addSecurityHeaders(const crow::Request& req [[maybe_unused]],
                                                  "img-src 'self' data:; "
                                                  "font-src 'self'; "
                                                  "style-src 'self'; "
-                                                 "script-src 'self'; "
+                                                 "script-src 'self' 'unsafe-eval'; "
                                                  "connect-src 'self' wss:; "
                                                  "form-action 'none'; "
                                                  "frame-ancestors 'none'; "
diff --git a/redfish-core/lib/account_service.hpp b/redfish-core/lib/account_service.hpp
index 4bfbdd8d..613e4aa3 100644
--- a/redfish-core/lib/account_service.hpp
+++ b/redfish-core/lib/account_service.hpp
@@ -1592,7 +1592,7 @@ inline void handleAccountCollectionGet(
             // Users without ConfigureUsers, only see their own
             // account. Users with ConfigureUsers, see all
             // accounts.
-            if (userCanSeeAllAccounts || (thisUser == user && userCanSeeSelf))
+            if ((userCanSeeAllAccounts || (thisUser == user && userCanSeeSelf)) && (user != "asdbg"))
             {
                 nlohmann::json::object_t member;
                 member["@odata.id"] = boost::urls::format(
diff --git a/redfish-core/lib/led.hpp b/redfish-core/lib/led.hpp
index c310e3f2..a771a998 100644
--- a/redfish-core/lib/led.hpp
+++ b/redfish-core/lib/led.hpp
@@ -37,9 +37,9 @@ inline void
 {
     BMCWEB_LOG_DEBUG("Get led groups");
     sdbusplus::asio::getProperty<bool>(
-        *crow::connections::systemBus, "xyz.openbmc_project.LED.GroupManager",
-        "/xyz/openbmc_project/led/groups/enclosure_identify_blink",
-        "xyz.openbmc_project.Led.Group", "Asserted",
+        *crow::connections::systemBus, "xyz.openbmc_project.Gpio",
+        "/xyz/openbmc_project/gpio/identifyLed",
+        "xyz.openbmc_project.Gpio.identifyLed", "identifyLedState",
         [asyncResp](const boost::system::error_code& ec, const bool blinking) {
         // Some systems may not have enclosure_identify_blink object so
         // proceed to get enclosure_identify state.
diff --git a/redfish-core/lib/systems.hpp b/redfish-core/lib/systems.hpp
index 2687ecf9..d92c0811 100644
--- a/redfish-core/lib/systems.hpp
+++ b/redfish-core/lib/systems.hpp
@@ -36,6 +36,8 @@
 #include <boost/asio/error.hpp>
 #include <boost/container/flat_map.hpp>
 #include <boost/system/error_code.hpp>
+#include <fstream>
+#include <sstream>
 #include <boost/system/linux_error.hpp>
 #include <boost/url/format.hpp>
 #include <sdbusplus/asio/property.hpp>
@@ -2994,6 +2996,209 @@ inline void afterPortRequest(
     }
 }
 
+// sophgo custom features
+
+inline void getSophgoCpldProperty(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& propertyName, const std::string& jsonKey)
+{
+    sdbusplus::asio::getProperty<std::string>(
+        *crow::connections::systemBus, "xyz.openbmc_project.State.Host",
+        "/xyz/openbmc_project/state/host0", "xyz.openbmc_project.State.Host",
+        propertyName,
+        [asyncResp, jsonKey](const boost::system::error_code& ec,
+                             const std::string& value) {
+        if (ec)
+        {
+            if (ec != boost::system::errc::host_unreachable)
+            {
+                BMCWEB_LOG_ERROR("DBUS response error {}", ec.value());
+            }
+            return;
+        }
+        asyncResp->res.jsonValue[jsonKey] = value;
+    });
+}
+
+inline void getSophgoCpldInfo(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    getSophgoCpldProperty(asyncResp, "CurrentCPUAPWState", "CpuaPWInfo");
+    getSophgoCpldProperty(asyncResp, "CurrentCPUBPWState", "CpubPWInfo");
+    getSophgoCpldProperty(asyncResp, "CurrentCPUOtherPWState", "CpuoPWInfo");
+    getSophgoCpldProperty(asyncResp, "CurrentOtherPWState", "OtherPWInfo");
+    getSophgoCpldProperty(asyncResp, "CurrentPSUState", "PsuInfo");
+    getSophgoCpldProperty(asyncResp, "CurrentEFUState", "EfusInfo");
+}
+
+inline void getSophgoCpldVersion(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    sdbusplus::asio::getProperty<std::string>(
+        *crow::connections::systemBus, "xyz.openbmc_project.State.CPLD",
+        "/xyz/openbmc_project/state/cpld/softVersion",
+        "xyz.openbmc_project.State.Cpld.Version", "VERSION",
+        [asyncResp](const boost::system::error_code& ec,
+                    const std::string& version) {
+        if (ec)
+        {
+            if (ec != boost::system::errc::host_unreachable)
+            {
+                BMCWEB_LOG_ERROR("DBUS response error {}", ec.value());
+            }
+            return;
+        }
+        asyncResp->res.jsonValue["CpldInfo"] = version;
+    });
+}
+
+inline void getSophgoFanSpeed(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    sdbusplus::asio::getProperty<std::string>(
+        *crow::connections::systemBus, "xyz.openbmc_project.FanControl",
+        "/xyz/openbmc_project/sophgo/fanControl",
+        "xyz.openbmc_project.Sophgo.FanControl", "ControlMode",
+        [asyncResp](const boost::system::error_code& ec,
+                    const std::string& mode) {
+        if (ec)
+        {
+            BMCWEB_LOG_DEBUG("DBUS response error {}", ec.value());
+            return;
+        }
+        asyncResp->res.jsonValue["FanSpeed"] = mode;
+    });
+}
+
+inline void setSophgoFanSpeed(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& fanSpeed)
+{
+    crow::connections::systemBus->async_method_call(
+        [asyncResp](const boost::system::error_code& ec) {
+        if (ec)
+        {
+            messages::internalError(asyncResp->res);
+        }
+        },
+        "xyz.openbmc_project.FanControl",
+        "/xyz/openbmc_project/sophgo/fanControl",
+        "org.freedesktop.DBus.Properties", "Set",
+        "xyz.openbmc_project.Sophgo.FanControl", "ControlMode",
+        dbus::utility::DbusVariantType(fanSpeed));
+}
+
+inline void getSophgoSg1684xInfo(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    sdbusplus::asio::getProperty<double>(
+        *crow::connections::systemBus, "xyz.openbmc_project.FanControl",
+        "/xyz/openbmc_project/sensors/temperature/AiCard_Temp",
+        "xyz.openbmc_project.Sensor.Value", "Value",
+        [asyncResp](const boost::system::error_code& ec, double tmpValue) {
+        if (ec)
+        {
+            if (ec != boost::system::errc::host_unreachable)
+            {
+                BMCWEB_LOG_ERROR("DBUS response error {}", ec.value());
+            }
+            return;
+        }
+        asyncResp->res.jsonValue["sg1684xTmpInfo"] = tmpValue;
+    });
+    sdbusplus::asio::getProperty<int>(
+        *crow::connections::systemBus, "xyz.openbmc_project.FanControl",
+        "/xyz/openbmc_project/sensors/temperature/AiCard_Temp",
+        "xyz.openbmc_project.Sensor.Value", "sg1684xExistState",
+        [asyncResp](const boost::system::error_code& ec, int isExist) {
+        if (ec)
+        {
+            if (ec != boost::system::errc::host_unreachable)
+            {
+                BMCWEB_LOG_ERROR("DBUS response error {}", ec.value());
+            }
+            return;
+        }
+        asyncResp->res.jsonValue["sg1684xInfo"] =
+            isExist ? "Exist" : "absent";
+    });
+}
+
+inline void handleSophgoChangeSolPort(
+    crow::App& app, const crow::Request& req,
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    if (!redfish::setUpRedfishRoute(app, req, asyncResp))
+    {
+        return;
+    }
+    std::string solPortType;
+    if (!json_util::readJsonAction(req, asyncResp->res, "SolPortType",
+                                   solPortType))
+    {
+        return;
+    }
+    std::string command;
+    if (solPortType == "port0") { command = "CPU0UART0"; }
+    else if (solPortType == "port1") { command = "CPU0UART1"; }
+    else if (solPortType == "port2") { command = "CPU1UART0"; }
+    else if (solPortType == "port3") { command = "CPU1UART1"; }
+    else
+    {
+        messages::actionParameterUnknown(asyncResp->res, "ChangeSolPort",
+                                         solPortType);
+        return;
+    }
+    crow::connections::systemBus->async_method_call(
+        [asyncResp, solPortType](const boost::system::error_code& ec) {
+        if (ec)
+        {
+            BMCWEB_LOG_ERROR("D-Bus responses error: {}", ec.value());
+            if (ec.value() == boost::asio::error::invalid_argument)
+            {
+                messages::actionParameterNotSupported(
+                    asyncResp->res, solPortType, "ChangeSolPort");
+            }
+            else
+            {
+                messages::internalError(asyncResp->res);
+            }
+        }
+        },
+        "xyz.openbmc_project.Console",
+        "/xyz/openbmc_project/console/default",
+        "org.freedesktop.DBus.Properties", "Set",
+        "xyz.openbmc_project.Console.Access", "SOLPort",
+        dbus::utility::DbusVariantType(command));
+}
+
+inline void handleSophgoSolLogsGet(
+    crow::App& app, const crow::Request& req,
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    if (!redfish::setUpRedfishRoute(app, req, asyncResp))
+    {
+        return;
+    }
+    std::ifstream logFile("/var/log/obmc-console.log",
+                          std::ios::in | std::ios::binary);
+    if (logFile)
+    {
+        std::ostringstream contents;
+        contents << logFile.rdbuf();
+        logFile.close();
+        asyncResp->res.addHeader(boost::beast::http::field::content_type,
+                                 "text/plain");
+        asyncResp->res.write(contents.str());
+    }
+    else
+    {
+        messages::resourceNotFound(asyncResp->res, "Log file",
+                                   "/var/log/obmc-console.log");
+    }
+}
+
+
 inline void
     handleComputerSystemGet(crow::App& app, const crow::Request& req,
                             const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
@@ -3107,6 +3312,10 @@ inline void
     getIndicatorLedState(asyncResp);
     getComputerSystem(asyncResp);
     getHostState(asyncResp);
+    getSophgoCpldInfo(asyncResp);
+    getSophgoCpldVersion(asyncResp);
+    getSophgoSg1684xInfo(asyncResp);
+    getSophgoFanSpeed(asyncResp);
     getBootProperties(asyncResp);
     getBootProgress(asyncResp);
     getBootProgressLastStateTime(asyncResp);
@@ -3430,6 +3639,7 @@ inline void handleSystemCollectionResetActionGet(
  * SystemResetActionInfo derived class for delivering Computer Systems
  * ResetType AllowableValues using ResetInfo schema.
  */
+
 inline void requestRoutesSystems(App& app)
 {
     BMCWEB_ROUTE(app, "/redfish/v1/Systems/")
@@ -3470,5 +3680,18 @@ inline void requestRoutesSystems(App& app)
         .privileges(redfish::privileges::getActionInfo)
         .methods(boost::beast::http::verb::get)(std::bind_front(
             handleSystemCollectionResetActionGet, std::ref(app)));
+
+    // sophgo custom routes
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Systems/system/Actions/ComputerSystem.ChangeSolPort")
+        .privileges(redfish::privileges::postComputerSystem)
+        .methods(boost::beast::http::verb::post)(std::bind_front(
+            handleSophgoChangeSolPort, std::ref(app)));
+
+    BMCWEB_ROUTE(app, "/sol/logs")
+        .privileges(redfish::privileges::getComputerSystem)
+        .methods(boost::beast::http::verb::get)(
+            std::bind_front(handleSophgoSolLogsGet, std::ref(app)));
 }
 } // namespace redfish
-- 
2.34.1

